File: src/types/contracts.ts
export type AlgorithmId = 'avsegformer' | 'avis' | 'vct' | 'combo'
export type InferenceScene = 'single_source' | 'multi_source'
export type TaskStatus = 'queued' | 'running' | 'completed' | 'failed' | 'canceled'
export type UserRole = 'admin' | 'user'

export interface UploadResponse {
  file_id: string
  filename: string
  size_bytes: number
  duration_seconds?: number
  width?: number
  height?: number
  fps?: number
  total_frames?: number
  audio_energy?: number
}

export interface CreateTaskRequest {
  file_id: string
  algorithm: AlgorithmId
  scene?: InferenceScene
}

export interface TaskProgress {
  task_id: string
  status: TaskStatus
  progress: number
  current_frame?: number
  total_frames?: number
  message?: string
  algorithm?: AlgorithmId
  scene?: InferenceScene
  resolved_scene?: 'single_source' | 'multi_source'
  filename?: string
  fps?: number
  duration_seconds?: number
  width?: number
  height?: number
  metrics?: TaskMetrics
  created_at?: string
  updated_at?: string
}

export interface TaskMetrics {
  jaccard?: number
  f_measure?: number
  jf_mean?: number
  map?: number
  hota?: number
  fsla?: number
  total_inference_ms?: number
  avg_frame_ms?: number
  processed_frames?: number
}

export interface TaskReport {
  task_id: string
  algorithm: AlgorithmId | string
  subset?: string
  frames?: number
  fps?: number
  duration_seconds?: number
  width?: number
  height?: number
  metrics?: TaskMetrics
  processing?: {
    total_ms?: number
    avg_frame_ms?: number
    processed_frames?: number
  }
  mask_coverage_pct_by_frame?: number[]
  note?: string
}

export interface AlgorithmInfo {
  id: AlgorithmId
  name: string
  version?: string
  description: string
  input_size?: string
  enabled: boolean
}

export interface AdminLoginResponse {
  token: string
  expires_at: string
  role: UserRole
}

export interface UserLoginResponse {
  token: string
  expires_at: string
  role: UserRole
}

export interface LogEntry {
  ts: string
  level: string
  message: string
}

File: api/schemas/contracts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


AlgorithmId = Literal["avsegformer", "avis", "vct", "combo"]
SceneId = Literal["single_source", "multi_source", "auto_detect"]
TaskStatus = Literal["queued", "running", "completed", "failed", "canceled"]


class UploadResponse(BaseModel):
    file_id: str
    filename: str
    size_bytes: int
    duration_seconds: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    fps: Optional[float] = None
    total_frames: Optional[int] = None
    audio_energy: Optional[float] = None
    recommended_scene: Optional[Literal["single_source", "multi_source"]] = None


class TaskMetrics(BaseModel):
    jaccard: Optional[float] = None
    f_measure: Optional[float] = None
    jf_mean: Optional[float] = None
    total_inference_ms: Optional[int] = None
    avg_frame_ms: Optional[float] = None
    processed_frames: Optional[int] = None


class CreateTaskRequest(BaseModel):
    file_id: str
    algorithm: AlgorithmId
    scene: Optional[SceneId] = None


class CreateTaskResponse(BaseModel):
    task_id: str


class TaskProgress(BaseModel):
    task_id: str
    status: TaskStatus
    progress: int = Field(ge=0, le=100)
    current_frame: Optional[int] = None
    total_frames: Optional[int] = None
    message: Optional[str] = None
    algorithm: Optional[AlgorithmId] = None
    scene: Optional[SceneId] = None
    resolved_scene: Optional[Literal["single_source", "multi_source"]] = None
    filename: Optional[str] = None
    fps: Optional[float] = None
    duration_seconds: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    metrics: Optional[TaskMetrics] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class AlgorithmInfo(BaseModel):
    id: str
    name: str
    version: Optional[str] = None
    description: str
    input_size: Optional[str] = None
    enabled: bool


class AdminLoginRequest(BaseModel):
    username: str
    password: str


class AdminLoginResponse(BaseModel):
    token: str
    expires_at: str
    role: Literal["admin", "user"]


class UserRegisterRequest(BaseModel):
    username: str
    password: str


class UserLoginRequest(BaseModel):
    username: str
    password: str


class UserLoginResponse(BaseModel):
    token: str
    expires_at: str
    role: Literal["admin", "user"]


class LogEntry(BaseModel):
    ts: str
    level: str
    message: str

